File: 92. Reverse Linked List II/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    ListNode su = null;
    
    public ListNode reverseN(ListNode head, int s) {
        if (s == 1) {
            su = head.next;
            return head;
        }
        ListNode last = reverseN(head.next, s - 1);
        head.next.next = head;
        head.next = su;
        return last;
    }
    
    public ListNode reverseBetween(ListNode head, int left, int right) {
        if (left == 1) {
            return reverseN(head, right);
        }
        head.next = reverseBetween(head.next, left - 1, right - 1);
        return head;
    }
}

File: 92. Reverse Linked List II/Solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
ListNode* ta;

class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        ListNode* dummy = new ListNode(-1);
        dummy->next = head;
        auto p = dummy;
        for (int i = 0; i < left - 1; i++) p = p->next;
        ListNode* l = p->next;
        p->next = reverse(p->next, right - left);
        if (l) l->next = ta;
        return dummy->next;
    }
    
    ListNode* reverse(ListNode* root, int len)
    {
        if (len == 0) {
            ta = root->next;
            return root;
        }
        ListNode* se = root->next;
        root->next = NULL;
        ListNode* newHead = reverse(se, len - 1);
        se->next = root;
        return newHead;
    }
};